proposals/0000-existentials.rst

```diff
@@ -0,0 +1,504 @@
+   in the type.
+
+   A point expanded in the `paper`_ is that this implementation is *lazy*,
+   just like ``filter`` should be. I do not believe a lazy implementation
+   of ``filter`` over length-indexed vectors is possible in GHC's current
+   type system.
+
+#. Suppose we have a pretty-printer based around the following class::
+
+     class Pretty a where
+       ppr :: a -> Doc
+
+   We would naturally have ::
+
+     instance Pretty a => Pretty [a] where ...
+
+   Yet, if I have ``woz :: Woz`` and ``wiz :: Wiz`` (with instances for
+   both types), I cannot ``ppr [woz, wiz]``, because that creates a
+   heterogeneous list.
+
+   With this proposal, I could write ::
+
+     pprList :: [exists a. Pretty a /\ a] -> Doc
+     pprList = sep . map ppr
+
+   and then have ``pprList [woz, wiz]``.
+
+   It would be even better to have ``ppr [woz, wiz]``, but that seems
+   beyond the abilities of type inference at the moment.
+
+#. The refinement types of Liquid Haskell often look something like this::
+
+     plusNat :: { x :: Nat } -> { y :: Nat } -> { v :: Nat | v >= x && v >= y }
+
+   where the result type has a refinement making a claim about the result
+   of running the function.
+
+   It would amplify the power of Liquid Haskell to have its refinement types
+   interact with other type system features in Haskell. Accordingly, we might
+   want to represent the inputs as pi-types and the output as a sigma-type --
+   which is essentially the same as an existential. Here might be one rendering::
+
+     plusNat :: foreach (x :: Nat) (y :: Nat) -> exists (v :: Nat). Proof (v >= x && v >= y)
+
+   Yet we do not want to manually pack and unpack the existential in the
+   definition for ``plusNat`` -- and thus need the inference capabilities proposed
+   here.
+
+   Note that this proposal does not go "all the way" toward this encoding of
+   refinement types, in that we would not be able to write the type above with
+   this proposal. Nevertheless, the automatic inference of packing and unpacking
+   described here seems necessary if we are to integrate Liquid Haskell with the
+   rest of GHC's type system.
+
+Proposed Change Specification
+-----------------------------
+
+1. Introduce a new extension ``-XExistentialTypes``.
+
+#. With ``-XExistentialTypes``, ``exists`` is a keyword in both
+   types and terms.
+
+#. With ``-XExistentialTypes``, introduce a new type for existentials.
+
+   1. The grammar is modified as follows (baseline: GHC's parser)::
+
+        ctype → forall_telescope ctype
+              | context '=>' ctype
+              | exists_telescope ctype   -- NEW!
+              | ctype
+              | ...
+
+          -- just for comparison
+        forall_telescope → 'forall' tv_bndrs '.'
+                         | 'forall' tv_bndrs '->'
+
+        exists_telescope → 'exists' tv_bndrs '.'
+
+      An existential is a new form of type, not equal to any current form.
```

Comment:
I assume the new form is not allowed in GADT constructor declarations. Please specify explicitly.